devito/interpolators.py

```diff
@@ -0,0 +1,354 @@
+import sympy
+import numpy as np
+
+from abc import ABC, abstractmethod
+
+from devito.equation import Eq, Inc
+from devito.logger import warning
+from devito.symbolics import retrieve_function_carriers, indexify, INT
+from devito.tools import Evaluable, powerset, flatten, prod
+from devito.types.basic import Scalar
+from devito.types.dense import SubFunction
+from devito.types.dimension import ConditionalDimension, Dimension, DefaultDimension
+
+
+class UnevaluatedSparseOperation(Evaluable):
+
+    """
+    Represents an Injection or an Interpolation operation performed on a
+    SparseFunction. Evaluates to a list of Eq objects.
+    objects.
+
+    Parameters
+    ----------
+    interpolator : Interpolator
+        Interpolator object that will be used to evaluate the Operation
+    *args, **kwargs
+        The arguments passed to the corresponding method
+    """
+    subdomain = None
+
+    def __init__(self, interpolator, *args, **kwargs):
+        self.interpolator = interpolator
+        self._args = args
+        self._kwargs = kwargs
+
+    def __add__(self, other):
+        return flatten([self, other])
+
+    def __radd__(self, other):
+        return flatten([other, self])
+
+
+class Injection(UnevaluatedSparseOperation):
+
+    """
+    Represents an Injection operation performed on a SparseFunction.
+    Evaluates to a list of Eq objects.
+    """
+
+    def __init__(self, interpolator, field, expr, *args, **kwargs):
+        self.interpolator = interpolator
+        self.field = field
+        self.expr = expr
+        self._args = args
+        self._kwargs = kwargs
+
+    @property
+    def evaluate(self):
+        return_value = self.interpolator.inject(self.field, self.expr, *self._args,
+                                                **self._kwargs)
+        assert(all(isinstance(i, Eq) for i in return_value))
+        return return_value
+
+    def __repr__(self):
+        return "Injection(%s on %s)" % (repr(self.expr), repr(self.field))
+
+
+class Interpolation(UnevaluatedSparseOperation):
+
+    """
+    Represents an Injection operation performed on a SparseFunction.
+    Evaluates to a list of Eq objects.
+    """
+
+    def __init__(self, interpolator, expr, *args, **kwargs):
+        self.interpolator = interpolator
+        self.expr = expr
+        self._args = args
+        self._kwargs = kwargs
+
+    @property
+    def evaluate(self):
+        return_value = self.interpolator.interpolate(self.expr, *self._args,
+                                                     **self._kwargs)
+        assert(all(isinstance(i, Eq) for i in return_value))
+        return return_value
+
+    def __repr__(self):
+        return "Interpolation(%s)" % repr(self.expr)
```

Comment:
So Interpolation doesn't know what it's interpolated into? (rec)

Reply:
Right now, directly, it doesn't.

Reply:
Fixed, but perhaps not in the way you wanted. Mark as resolved if you don't mind my fix @mloubout .